invokeai/backend/stable_diffusion/diffusion/custom_attention.py

```diff
@@ -0,0 +1,432 @@
+            # (batch, heads, source_length, target_length)
+            attention_mask = attention_mask.view(batch_size, attn.heads, -1, attention_mask.shape[-1])
+
+        # the output of sdp = (batch, num_heads, seq_len, head_dim)
+        hidden_states = F.scaled_dot_product_attention(
+            query, key, value, attn_mask=attention_mask, dropout_p=0.0, is_causal=False, scale=attn.scale
+        )
+
+        hidden_states = hidden_states.transpose(1, 2).reshape(batch_size, -1, attn.heads * head_dim)
+        hidden_states = hidden_states.to(query.dtype)
+
+        return hidden_states
+
+    def run_attention_sliced(
+        self,
+        attn: Attention,
+        query: torch.Tensor,
+        key: torch.Tensor,
+        value: torch.Tensor,
+        attention_mask: Optional[torch.Tensor],
+        slice_size: int,
+    ) -> torch.Tensor:
+        dim = query.shape[-1]
+
+        query = attn.head_to_batch_dim(query)
+        key = attn.head_to_batch_dim(key)
+        value = attn.head_to_batch_dim(value)
+
+        batch_size_attention, query_tokens, _ = query.shape
+        hidden_states = torch.zeros(
+            (batch_size_attention, query_tokens, dim // attn.heads), device=query.device, dtype=query.dtype
+        )
+
+        for i in range((batch_size_attention - 1) // slice_size + 1):
+            start_idx = i * slice_size
+            end_idx = (i + 1) * slice_size
+
+            query_slice = query[start_idx:end_idx]
+            key_slice = key[start_idx:end_idx]
+            value_slice = value[start_idx:end_idx]
+            attn_mask_slice = attention_mask[start_idx:end_idx] if attention_mask is not None else None
+
+            if self.attention_type == "normal":
+                attn_slice = attn.get_attention_scores(query_slice, key_slice, attn_mask_slice)
+                torch.bmm(attn_slice, value_slice, out=hidden_states[start_idx:end_idx])
+                del attn_slice
+            elif self.attention_type == "xformers":
+                if attn_mask_slice is not None:
+                    attn_mask_slice = attn_mask_slice.expand(-1, query.shape[1], -1)
+
+                hidden_states[start_idx:end_idx] = xformers.ops.memory_efficient_attention(
+                    query_slice, key_slice, value_slice, attn_bias=attn_mask_slice, op=None, scale=attn.scale
+                )
+            elif self.attention_type == "torch-sdp":
+                if attn_mask_slice is not None:
+                    attn_mask_slice = attn_mask_slice.unsqueeze(0)
+
+                hidden_states[start_idx:end_idx] = F.scaled_dot_product_attention(
+                    query_slice.unsqueeze(0),
+                    key_slice.unsqueeze(0),
+                    value_slice.unsqueeze(0),
+                    attn_mask=attn_mask_slice,
+                    dropout_p=0.0,
+                    is_causal=False,
+                    scale=attn.scale,
+                ).squeeze(0)
```

Comment:
Don't mean to keep adding comments, but just happened to notice this. I'm pretty sure all this squeezing/unsqueezing is unnecessary:

```suggestion
                hidden_states[start_idx:end_idx] = F.scaled_dot_product_attention(
                    query_slice,
                    key_slice,
                    value_slice,
                    attn_mask=attn_mask_slice,
                    dropout_p=0.0,
                    is_causal=False,
                    scale=attn.scale,
                )
```

Reply:
Sometimes it gives even 2x performance boost on cuda.

Reply:
Redone code, this time reimplemented it, not reorganized diffusers' one and added some comments.